### PR TITLE
CORDA-1477 add check for code version in checkpoints

### PR DESCRIPTION
**Pull request by tudor-malene**
https://r3-cev.atlassian.net/browse/CORDA-1477

The general idea is that with each checkpoint there is a SubFlow stack.
For each SubFlow I've added a "SubFlowVersion" which can be either a CorDapp jar hash or the Core platform version.

On node startup, each checkpoint is deserialized, and the Subflow versions are checked against what is currently installed ( CheckpointStorage.verifyCheckpointsCompatible)

What is required is an Integration test that will do the following:
- create network with 2 nodes (NodeA, NodeB)
- run some flows , and the stop NodeB and run more flows ( just to have some checkpoints)
- on NodeA deploy a new version of the CorDapp ( not sure how..)
- try to start NodeA, it should fail
- revert to the old Cordapp, start nodeB, clear the checkpoints
- nodeA should start successfully

@szymonsztuka ^^
For me, this sounds like a test that is suitable for the behave format.

### CODE REVIEW

**Comment by exFalso** [comment]
Another thing I just realised is that this feature requires backwards compatibility wrt `Checkpoint` serialization. This does not involve the user's part of the checkpoint (the flow data), but it does involve the `Checkpoint` datastructure itself.

For example this PR modifies `SubFlow`, so if we try to deserialize a `Checkpoint` created before this PR then we won't reach the version check, it will just fail to deserialize.

`Checkpoint`s are serialized by kryo, so we need to explore what sort of forwards/backwards compatibility they have. Alternatively we can switch to AMQP just for the `Checkpoint` datastructure, and later use kryo for the flow data (`frozenFiber`/`frozenFlowLogic`)

**Comment by exFalso** [comment]
There is another subtlety here. `frozenFiber` and `frozenFlowLogic` doesn't *only* contain flow data, it may involve some (although minimal) platform specific data the version of which is *not* pinned by the cordapp hashes. An example is `FlowStateMachineImpl.id` or `FlowLogic.deprecatedPartySessionMap`.

This to me means we may want to incorporate the platform version check for every checkpoint, regardless of whether there is a platform subflow in it.

@mikehearn what are your thoughts on the above?

**Comment by tudor-malene** [comment]
Good points @exFalso !
Regarding deserialising old checkpoints, can we register a "smart" customer deserialiser for SubFlow?

**Comment by tudor-malene** [comment]
@mikehearn , @exFalso 

Just realised there is another problem.
Let's say you're on Corda v7, and have some checkpoints in the db.
Suddenly you decide to upgrade to Corda v8 without draining.

Let's say you do the usual steps of migrating the db.
When you start the node it will complain that the checkpoints are incompatible, but to revert to the old version of the node, you need to rollback the db to the previous version.

I wonder, should we integrate a check when we run db migration that the checkpoints table needs to be empty?

**Comment by exFalso** [comment]
@tudor-malene That made me think, if we had such a check during database migration, or in general if we enforced that a *platform* upgrade can only happen with an empty checkpoint table then this also solves the compatibility issues wrt the `Checkpoint` datastructure!

**Comment by tudor-malene** [comment]
Back to you @exFalso .
I will add the db migration check to ENT after the merge

File: node/src/main/kotlin/net/corda/node/internal/CheckpointVerifier.kt [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+/**
+ * Thrown at startup, if a checkpoint is found that is incompatible with the current code
+ */
+sealed class CheckpointIncompatibleException(override val message: String) : Exception() {
+    class CannotBeDeserialisedException(val e: Exception) : CheckpointIncompatibleException(
+            "Found checkpoint that cannot be deserialised using the current Corda version. Please revert to the previous version of Corda, drain your node, and try again. Cause: ${e.message}")
+
+    class SubFlowCoreVersionIncompatibleException(val flowClass: Class<out FlowLogic<*>>, oldVersion: Int) : CheckpointIncompatibleException(
+            "Found checkpoint for flow: ${flowClass} that is incompatible with the current Corda platform. Please revert to the previous version of Corda (v = ${oldVersion}), drain your node, and try again.")
```
Comment (tlil) [inline node/src/main/kotlin/net/corda/node/internal/CheckpointVerifier.kt]: `v = ${oldVersion}` => `version ${oldVersion}`

File: node/src/main/kotlin/net/corda/node/internal/CheckpointVerifier.kt [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        }
+    }
+}
+
+/**
+ * Thrown at startup, if a checkpoint is found that is incompatible with the current code
+ */
+sealed class CheckpointIncompatibleException(override val message: String) : Exception() {
+    class CannotBeDeserialisedException(val e: Exception) : CheckpointIncompatibleException(
+            "Found checkpoint that cannot be deserialised using the current Corda version. Please revert to the previous version of Corda, drain your node, and try again. Cause: ${e.message}")
```
Comment (tlil) [inline node/src/main/kotlin/net/corda/node/internal/CheckpointVerifier.kt]: When referring to draining your node, perhaps even link to the doc site as we do elsewhere? There's some information on https://docs.corda.net/upgrading-cordapps.html#flow-drains

**Comment by tudor-malene** [comment]
Just realised another issue. When devMode=true, should we relax this and just print a warning?
As a dev, you might be working on a flow, and this could stand in your way.

**Comment by tudor-malene** [comment]
Ready for re-review @exFalso .

After adding the check that there should be only 1 CorDapp providing any Flow, I broke some tests.
Turned out, we were dynamically adding core and finance jars twice to the driver nodes.
I tweaked that a bit.

**Pull request merged**